spec/alephant/logger/json_spec.rb

```diff
@@ -1,137 +1,92 @@
+    allow(File).to receive(:open) { log_output_obj }
+    allow(log_output_obj).to receive :sync=
   end
 
-  shared_examples 'JSON logging' do
-    let(:log_hash) do
-      { 'foo' => 'bar', 'baz' => 'quux' }
-    end
-
-    it 'writes JSON dump of hash to log with corresponding level key' do
-      allow(Time).to receive(:now).and_return('foobar')
+  logging_levels = Alephant::Logger::LEVELS.map(&:to_s)
+  levels_size = logging_levels.size - 1
 
-      expect(log_file).to receive(:write) do |json_dump|
-        h = { 'timestamp' => 'foobar', 'uuid' => 'n/a', 'level' => level }
-        expect(JSON.parse(json_dump)).to eq h.merge log_hash
-      end
-
-      subject.send(level, log_hash)
-    end
-
-    it 'automatically includes a timestamp' do
-      expect(log_file).to receive(:write) do |json_dump|
-        t = JSON.parse(json_dump)['timestamp']
-        expect { DateTime.parse(t) }.to_not raise_error
-      end
-
-      subject.send(level, log_hash)
-    end
+  %w(debug info warn error).each_with_index do |level, i|
+    describe "##{level}" do
+      let(:level) { level }
+      let(:log_hash) { { 'foo' => 'bar', 'baz' => 'quux' } }
 
-    it 'outputs the timestamp first' do
-      expect(log_file).to receive(:write) do |json_dump|
-        h = JSON.parse(json_dump)
-        expect(h.first[0].to_sym).to be :timestamp
-      end
+      context 'when the desired write level is specified' do
+        subject(:logger) do
+          described_class.new(log_path, level: desired_write_level)
+        end
 
-      subject.send(level, log_hash)
-    end
+        context 'when same as the desired write level' do
+          let(:desired_write_level) { level.to_sym }
 
-    it 'displays a default session value if a custom function is not provided' do
-      expect(log_file).to receive(:write) do |json_dump|
-        h = JSON.parse(json_dump)
-        expect(h['uuid']).to eq 'n/a'
-      end
+          it_behaves_like 'a JSON log writer'
 
-      subject.send(level, log_hash)
-    end
+          it_behaves_like 'nests flattened to strings'
 
-    it 'displays a custom session value when provided a user defined function' do
-      expect(log_file).to receive(:write) do |json_dump|
-        h = JSON.parse(json_dump)
-        expect(h['uuid']).to eq 'foo'
-      end
+          it_behaves_like 'gracefully fails with string arg'
+        end
 
-      described_class.session = fn
-      subject.send(level, binding, log_hash)
-      described_class.session = -> { 'n/a' }
-    end
+        context 'when lower than the desired write level' do
+          let(:desired_write_level) do
+            idx = logging_levels.index(level)
+            logging_levels[i == levels_size ? i : idx + 1].to_sym
+          end
 
-    it 'provides a static method for checking if a session has been set' do
-      described_class.session = fn
-      expect(described_class.session?).to eq 'instance-variable'
+          it_behaves_like 'a JSON log non writer' unless i == levels_size
```

Comment:
I don't think we need the unless in this test. As we are in the context of "when lower than the desired write level" so we know the message level is lower than the desired. I don't see what the unless gives us. Also in tests, I think any logic could provide tests that can sometimes pass/fail in certain conditions.

Reply:
@samfrench As the tests are run in an each block. This prevents the example from being run on the last iteration where the "levels" would be the same; whereby the example would fail. I'll look into how we can remove it.